mne/gui/_core.py

```diff
@@ -0,0 +1,483 @@
+# -*- coding: utf-8 -*-
+"""Shared GUI classes and functions."""
+
+# Authors: Alex Rockhill <aprockhill@mailbox.org>
+#
+# License: BSD (3-clause)
+
+import os
+import os.path as op
+import numpy as np
+from functools import partial
+
+from qtpy import QtCore
+from qtpy.QtCore import Slot
+from qtpy.QtWidgets import (QMainWindow, QGridLayout,
+                            QVBoxLayout, QHBoxLayout, QLabel,
+                            QMessageBox, QWidget, QPlainTextEdit)
+
+from matplotlib import patheffects
+from matplotlib.backends.backend_qt5agg import FigureCanvas
+from matplotlib.figure import Figure
+from matplotlib.patches import Rectangle
+
+from .._freesurfer import _import_nibabel
+from ..viz.backends.renderer import _get_renderer
+from ..viz.utils import safe_event
+from ..surface import _read_mri_surface, _marching_cubes
+from ..transforms import apply_trans, _frame_to_str
+from ..utils import logger, _check_fname, verbose, warn, get_subjects_dir
+
+_IMG_LABELS = [['I', 'P'], ['I', 'L'], ['P', 'L']]
+_ZOOM_STEP_SIZE = 5
+
+
+@verbose
+def _load_image(img, verbose=None):
+    """Load data from a 3D image file (e.g. CT, MR)."""
+    nib = _import_nibabel('use GUI')
+    if not isinstance(img, nib.spatialimages.SpatialImage):
+        logger.info(f'Loading {img}')
+        _check_fname(img, overwrite='read', must_exist=True)
+        img = nib.load(img)
+    # get data
+    orig_data = np.array(img.dataobj).astype(np.float32)
+    # reorient data to RAS
+    ornt = nib.orientations.axcodes2ornt(
+        nib.orientations.aff2axcodes(img.affine)).astype(int)
+    ras_ornt = nib.orientations.axcodes2ornt('RAS')
+    ornt_trans = nib.orientations.ornt_transform(ornt, ras_ornt)
+    img_data = nib.orientations.apply_orientation(orig_data, ornt_trans)
+    orig_mgh = nib.MGHImage(orig_data, img.affine)
+    aff_trans = nib.orientations.inv_ornt_aff(ornt_trans, img.shape)
+    vox_ras_t = np.dot(orig_mgh.header.get_vox2ras_tkr(), aff_trans)
+    return img_data, vox_ras_t
+
+
+def _make_slice_plot(width=4, height=4, dpi=300):
+    fig = Figure(figsize=(width, height), dpi=dpi)
+    canvas = FigureCanvas(fig)
+    ax = fig.subplots()
+    fig.subplots_adjust(bottom=0, left=0, right=1, top=1, wspace=0, hspace=0)
+    ax.set_facecolor('k')
+    # clean up excess plot text, invert
+    ax.invert_yaxis()
+    ax.set_xticks([])
+    ax.set_yticks([])
+    return canvas, fig
+
+
+class SliceBrowser(QMainWindow):
```

Comment:
More future compatible perhaps to make this a QWidget? Then it can be embedded in any main window, popup, etc. We could do this in a separate PR, though, if that makes more sense / keeps the diff smaller (which it might, since the iEEG GUI code now just mostly wraps to this)

Reply:
Hmmm this is a good point but then it won't pop up in a main window for its own tests. We could easily wrap it though. Conceptually, it does make sense as a widget. Do you mind if I push this to the next PR when I put it in the TFR source space browser? I think also, I've been working on the abstraction code and that might effect it so it might be better to address in parts.

Reply:
Yeah a follow-up PR is fine

> but then it won't pop up in a main window for its own tests.

Have you tried? IIRC Qt might instantiate a window for you if you do this, and even if it doesn't actually pop up, I think that's okay for tests anyway. We just need to make sure the interactions / signals etc. actually work, which we do non-visually anyway

Reply:
... okay at least creating some simple widgets it did not show. But that might end up being okay in testing. If it's not, I agree some simple wrapper (probably a pytest fixture that returns a window -- pytest-qt might even have this already?) would make sense.